Product implementation

Osnaga-Robert opened the pull request:
- Product entity
- Category entity
- implementation of adding an image
- implementation of adding a product
- implementation of adding a category by the admin
- implementation of exception handler

Osnaga-Robert committed:
Product

Adding a product

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/entities/Image.java:
```diff
@@ -0,0 +1,27 @@
+    @Column(length = 5000000)
+    private byte[] imageByte;
+    
+    public Image() {}
```
You can get rid of this line of code (line 20) by adding @NoArgsConstructor annotation above the class.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/services/UserService.java:
```diff
@@ -19,15 +21,18 @@ public class UserService {
 
     //seller registration
     public User registerNewSeller(User seller) {
+        validateAllCredentialsSeller(seller);
         validateEmail(seller.getUserEmail());
         validatePassword(seller.getUserPassword());
```
I think you can do the following:
1) Rename validateAllCredentialsSellers method with validateCredentials.
2) The parameter for this method will be some Generic Type (T). Depending of the type of object (Seller or Buyer) do the specific logic for you entity.
3) You can also add the validateEmail() and validatePassword into the validateCredentials method.

This way you can avoid the code duplication.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/services/UserService.java:
```diff
@@ -76,36 +82,64 @@ public void initRolesandUser() {
         if (password.length() < 8) {
-            throw new RequestException("Password must be at least 8 characters long");
+            throw new RegisterException("Password must be at least 8 characters long");
         }
         if (!password.matches(".*[A-Z].*")) {
-            throw new RequestException("Password must contain at least one uppercase letter");
+            throw new RegisterException("Password must contain at least one uppercase letter");
         }
         if (!password.matches(".*[a-z].*")) {
-            throw new RequestException("Password must contain at least one lowercase letter");
+            throw new RegisterException("Password must contain at least one lowercase letter");
         }
         if (!password.matches(".*\\d.*")) {
-            throw new RequestException("Password must contain at least one digit");
+            throw new RegisterException("Password must contain at least one digit");
         }
         if (!password.matches(".*[!@#$%^&*].*")) {
-            throw new RequestException("Password must contain at least one special character (!@#$%^&*)");
+            throw new RegisterException("Password must contain at least one special character (!@#$%^&*)");
+        }
+    }
```
Instead of having 4 ifs with some parts of a pattern, can you create a constant variable with all those parts of a pattern in a single big pattern that will be served as a Regex pattern for your desired password format and then check if the password matches the pattern and handle the potential exceptions? You can do something similar to the logic from the validateEmail() method.

PatrickDiallo23 commented on shop4All-backend/src/main/java/com/example/shop4All_backend/services/CategoryService.java:
```diff
@@ -0,0 +1,52 @@
+@Service
+public class CategoryService {
+
+    @Autowired
+    private CategoryRepo categoryRepo;
```
I think it is better to make this variable final and then use @RequiredArgsConstructor annotation above the CategoryService class because you don't want to reassign the categoryRepo variable.

PatrickDiallo23 reviewed (changes requested):
Look at the comments, please.

Osnaga-Robert committed:
Updates

Updated products and users functionalities

Osnaga-Robert commented:
- added @NoArgsConstructor annotation at Image.java
- modified the validation of credentials from UserService.java
- modified exception password
- added @requiredargsconstructor for CategoryService.java

PatrickDiallo23 reviewed (approved):
Looks good to me!
